src/grain/ledger.js

```diff
@@ -0,0 +1,161 @@
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+export type Distribution = {|
+  +type: "DISTRIBUTION",
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+export type AliasSpec = {|
+  +canonical: NodeAddressT,
+  +alias: NodeAddressT,
+|};
+
+export type LedgerHistory = {|
+  +events: $ReadOnlyArray<LedgerEvent>,
+  +aliases: Map<NodeAddressT, NodeAddressT>,
+|};
+
+export interface Ledger {
+  balances(): Map<NodeAddressT, Grain>;
+  earnings(): Map<NodeAddressT, Grain>;
+  events(EventsOptions): $ReadOnlyArray<LedgerEvent>;
+  history(): LedgerHistory;
+}
+
+// $ExpectFlowError
+const zero: Grain = 0n;
+
+export class InMemoryLedger implements Ledger {
+  _balances: Map<NodeAddressT, Grain>;
+  _earnings: Map<NodeAddressT, Grain>;
+  _history: LedgerHistory;
+
+  constructor(history: LedgerHistory) {
+    this._history = history;
+    this._balances = new Map();
+    this._earnings = new Map();
+    let lastTimestamp = -Infinity;
+    for (const e of history.events) {
+      if (e.timestampMs < lastTimestamp) {
+        throw new Error(
+          `event timestamps out of order: encountered ${e.timestampMs} after ${lastTimestamp}`
+        );
+      }
+      switch (e.type) {
+        case "DISTRIBUTION":
+          this._processDistribution((e: Distribution));
+          break;
+        case "TRANSFER":
+          this._processTransfer((e: Transfer));
+          break;
+        default:
+          throw new Error((e.type: empty));
+      }
+    }
+  }
+
+  events(opts: EventsOptions): $ReadOnlyArray<LedgerEvent> {
+    let filter = (_) => true;
+    if (opts.address) {
+      filter = (x: LedgerEvent) => {
+        switch (x.type) {
+          case "DISTRIBUTION":
+            return x.recipient === opts.address;
+          case "TRANSFER":
+            return x.recipient === opts.address || x.sender === opts.address;
+          default:
+            throw new Error((x.type: empty));
+        }
+      };
+    }
+    const events = this._history.events.filter(filter);
+    if (opts.order === Order.DESCENDING) {
+      events.reverse();
+    }
+    return events;
+  }
+
+  history() {
+    return this._history;
+  }
+
+  balances(): Map<NodeAddressT, Grain> {
+    return new Map(this._balances);
+  }
+
+  _balance(a: NodeAddressT): Grain {
+    return NullUtil.orElse(this._balances.get(a), zero);
+  }
+
+  _earning(a: NodeAddressT): Grain {
+    return NullUtil.orElse(this._earnings.get(a), zero);
+  }
+
+  earnings(): Map<NodeAddressT, Grain> {
+    return new Map(this._earnings);
+  }
+
+  canonicalAddress(a: NodeAddressT): NodeAddressT {
+    return NullUtil.orElse(this._history.aliases.get(a), a);
+  }
+
+  _processDistribution(d: Distribution) {
+    const recipient = this.canonicalAddress(d.recipient);
+    const balance = this._balance(recipient) + d.amount;
+    this._balances.set(recipient, balance);
+    const earned = NullUtil.orElse(this._earnings.get(recipient), zero);
```

Comment:
Nit: earnings also has this `get() || 0` helper function. Should use for consistency.

Reply:
Can't find it. I assume this comment is outdated.

Reply:
I mean
```js
  _earning(a: NodeAddressT): Grain {
    return NullUtil.orElse(this._earnings.get(a), zero);
  }
```

So this line can be `const earned = this._earning(recipient);`
Same as the balance above.

Reply:
Got it. Fixed.